### ISSUE TITLE
Question regarding SMC

### ISSUE THREAD

**Issue by JossMarley** [OP-TEE/optee_os#1443]
Hi, when a SMC is invoked by `thread_*.S` we have a switch from the secure world to the normal world. Could you please explain me which is piece of code in normal world that is executed first after a SMC is performed? What is the function that gets executed first? I suppose the SMC is not handled directly by the tee supplicant... Am I right? I can't figure out what happens in the normal world after the SMC.

**Comment by jenswi-linaro** [comment]
Secure world is entered by an SMC in normal world (typically from `optee_do_call_with_arg()`). When OP-TEE is doing an SMC the secure monitor switches back to normal world an resumes execution with the next instruction after the original SMC in normal world (typically `optee_do_call_with_arg()` sees it as `optee->invoke_fn()` is returning).

**Comment by JossMarley** [comment]
Thank you. So we can say that after the switch between normal and secure world and vice versa normally it is `optee_do_call_with_arg` the first function executed. Right?
In the documentation file `interrupt handling` in the section `Entry and exit of Trusted OS` there is a something called `TEE_FUNC_INVOKE` so we are saying that that function is `optee_do_call_with_arg`?
Last question, in that file there is also a reference on saving and restoring  secure/non-secure context. Where is the code do that?

**Comment by jenswi-linaro** [comment]
No, the switching is done in `core/arch/arm/sm/sm_a32.S` and friends.

**Comment by etienne-lms** [comment]
@JossMarley, just a comment about context switching: as Jens stated, `core/arch/arm/sm/` containt the so-called "secure monitor" that manages these secure/non-secure CPU context switches, but that is applicable only to ARMv7 architectures and ARMv8 executing in AArch32 only. 
For ARMv8 architectures, it is likely to be the ARM-TrustedFirmware that is in charge of the CPU context switching. If you are targeting an ARMv8 platform, you should look at the OP-TEE SPD in ARM-TF for [invocation requests](https://github.com/ARM-software/arm-trusted-firmware/blob/master/services/spd/opteed/opteed_main.c#L210) or [secure interrupt notifications](https://github.com/ARM-software/arm-trusted-firmware/blob/master/services/spd/opteed/opteed_main.c#L77).

Regarding non-secure world, note that the TEE (optee) is invoked on request of the REE (linux). Hence as Jens said, at some point, linux calls `optee_do_call_with_arg()` (linux optee driver) which issues a SMC instruction. The SMC exception is trapped by the "secure monitor" (optee "sm" or ARM-TF runtime service). When the secure world returns to linux, it branches back to the instruction that follows the SMC instruction in `optee_do_call_with_arg()` and `optee_do_call_with_arg()` sequence resumes (warning: this function is not called, it resumes its execution).

**Comment by JossMarley** [comment]
thank you @etienne-lms your answer is very helpful. Just another doubt: where secure/non-secure context is saved/restored after switch from normal to secure world and vice versa?
